### PR TITLE
[rpc] refactor and move createException function

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29605 [rpc] refactor and move createException function**

Adds a wrapper around the existing `createException` function that
allows passing of an error string, instead of a regular C++ exception. This
allows us to mark futures with exceptions for errors that aren't necessarily c++
exceptions. This function is used by
https://github.com/pytorch/pytorch/pull/29601 and https://github.com/pytorch/pytorch/pull/26336.

Differential Revision: [D18439216](https://our.internmc.facebook.com/intern/diff/D18439216/)

### CODE REVIEW

**Review by zhaojuanmao — APPROVED** [review]
lgtm! please land it after tests are green

File: torch/csrc/distributed/rpc/message.h [hunk 1]
```diff
@@ -115,6 +115,14 @@ class TORCH_API Message final {
   int64_t id_ = -1;
 };
 
+// create an exception given a message and exception.
+TORCH_API Message
+createException(const Message& request, const std::exception& e);
```
Comment (pritamdamania87) [inline torch/csrc/distributed/rpc/message.h]: nit: Maybe we should call this `createExceptionResponse` and clearly state in the docs that we're building a `response` Message with an exception for the provided `request` Message.

File: torch/csrc/distributed/rpc/message.cpp [hunk 2]
```diff
@@ -113,6 +113,24 @@ void Message::setId(int64_t id) {
   id_ = id;
 }
 
+Message createExceptionResponse(
+    const Message& request,
+    const std::exception& e) {
+  std::string exceptionMsg = e.what();
+  return createExceptionResponse(request, exceptionMsg);
```
Comment (pietern) [inline torch/csrc/distributed/rpc/message.cpp]: Can you use `e.what()` inline?

File: torch/csrc/distributed/rpc/message.h [hunk 3]
```diff
@@ -115,6 +115,17 @@ class TORCH_API Message final {
...
 };
 
+// Create a response Message with an exception for the provided request message.
+// The exception string representation will be used as the message's payload.
+TORCH_API Message
+createExceptionResponse(const Message& request, const std::exception& e);
+
+// Create a response Message with an exception type for the provided request
+// message. The passed in string will be used as the created message's payload
+TORCH_API Message createExceptionResponse(
```
Comment (pritamdamania87) [inline torch/csrc/distributed/rpc/message.h]: It doesn't look like this method is used directly. Is there a reason we have this method? If not, can we remove it and only have one `createExceptionResponse` method for now?

**Review by pritamdamania87 — APPROVED** [review]

File: torch/csrc/distributed/rpc/message.h (hunk 3 above)
Comment (rohan-varma) [reply torch/csrc/distributed/rpc/message.h]: I am planning on using them in the 2 PRs I have linked in the description, so I just added it here to remove it from the other 2.

**Comment by rohan-varma** [comment]
The CI failures are unrelated, will proceed with landing this

**Pull request closed without merging**